pkg/lease/manager.go

```diff
@@ -0,0 +1,205 @@
+package lease
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	coordv1 "k8s.io/api/coordination/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
+	UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error)
+	InvalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error
+}
+
+type manager struct {
+	client.Client
+	log logr.Logger
+}
+
+func (l *manager) CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error) {
+	return l.createOrGetExistingLease(ctx, node, duration, holderIdentity, namespace)
+}
+
+func (l *manager) UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error) {
+	return l.updateLease(ctx, node, lease, currentTime, leaseDuration, leaseDeadline, holderIdentity)
+}
+
+func (l *manager) InvalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error {
+	return l.invalidateLease(ctx, nodeName, leaseNamespace)
+}
+
+func NewManager(cl client.Client) Manager {
+	return NewManagerWithCustomLogger(cl, ctrl.Log.WithName("leaseManager"))
+
+}
+
+func NewManagerWithCustomLogger(cl client.Client, log logr.Logger) Manager {
+	return &manager{
+		Client: cl,
+		log:    log,
+	}
+}
+
+func (l *manager) createOrGetExistingLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, leaseNamespace string) (*coordv1.Lease, bool, error) {
+	owner := makeExpectedOwnerOfLease(node)
+	microTimeNow := metav1.NowMicro()
+
+	lease := &coordv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
```

Comment:
still relevant: since this can be used for more things than just nodes, we can run into conflicts when creating leases of different resources but same name. So my suggestion is to prefix the name with the kind of the resource

Reply:
still unresolved

Reply:
Hmm maybe I misunderstood your intention, isn't [this](https://github.com/medik8s/common/pull/1/commits/310da66a49ddb25e6fcc6043f3992a2268b2c8e6)
what you meant by that comment ?